compiler/src/dotty/tools/dotc/typer/Inferencing.scala

```diff
@@ -162,6 +164,60 @@ object Inferencing {
       )
   }
 
+  def approximateGADT(tp: Type)(implicit ctx: Context): Type = {
+    val map = new IsFullyDefinedAccumulator2
+    val res = map(tp)
+    assert(!map.failed)
+    debug.println(i"approximateGADT( $tp )  = $res  //  {${tp.toString}}")
+    res
+  }
+
+  private class IsFullyDefinedAccumulator2(implicit ctx: Context) extends TypeMap {
+
+    var failed = false
+
+    private def instantiate(tvar: TypeVar, fromBelow: Boolean): Type = {
+      val inst = tvar.instantiate(fromBelow)
+      typr.println(i"forced instantiation of ${tvar.origin} = $inst")
+      inst
+    }
+
+    private def instDirection2(sym: Symbol)(implicit ctx: Context): Int = {
+      val constrained = ctx.gadt.fullBounds(sym)
+      val original = sym.info.bounds
+      val cmp = ctx.typeComparer
+      val approxBelow =
+        if (!cmp.isSubTypeWhenFrozen(constrained.lo, original.lo)) 1 else 0
+      val approxAbove =
+        if (!cmp.isSubTypeWhenFrozen(original.hi, constrained.hi)) 1 else 0
+      approxAbove - approxBelow
+    }
+
+    private[this] var toMaximize: Boolean = false
+
+    def apply(tp: Type): Type = tp.dealias match {
+      case tp @ TypeRef(qual, nme) if (qual eq NoPrefix) && ctx.gadt.contains(tp.symbol) =>
+        val sym = tp.symbol
+        val res =
+          ctx.gadt.approximation(sym, fromBelow = variance < 0)
+
+        debug.println(i"approximated $tp  ~~  $res")
```

Comment:
Overally feedback re. debug stmts (I know that most of these are leftovers from me :P):

1) My general impression is that using `debug` Printer is to be avoided in master-branch code. `debug` is mostly used for, as name suggests, quick debugging stmts and we usually want only the stmts we just added to print, not the ones half a codebase away. For this message, I'd use the `gadt` printer.
2) At a glance, there's too much debugging stmts remaining. As a first approximation, I'd say only this statement should be left, but since you were the one to work on this code most recently, I'll leave it to you to decide if other stmts should also be left, for the convenience of future debugging.

Reply:
Thanks, I have left this one and another one that displays some information from the GADT context, as I think it was most useful for debugging. I moved both to the `gadts` printer.